lua/nvim-lsp-ts-utils/inlay-hints.lua

```diff
@@ -2,99 +2,156 @@ local o = require("nvim-lsp-ts-utils.options")
 local u = require("nvim-lsp-ts-utils.utils")
 
 local api = vim.api
-
+local INLAY_HINTS_METHOD = "typescript/inlayHints"
 local M = {}
 
-M.state = {
-    enabled = false,
-    ns = api.nvim_create_namespace("ts-inlay-hints"),
-}
-
-function M.setup_autocommands()
-    u.buf_autocmd(
-        "TSLspImportOnCompletion",
-        "BufEnter,BufWinEnter,TabEnter,BufWritePost,TextChanged,TextChangedI",
-        "autocmd_fun()"
-    )
+local ns = api.nvim_create_namespace("ts-inlay-hints")
+M.ns = ns
+
+-- enabled[bufnr] = false
+M.enabled = {}
+
+local function buf_enabled(bufnr)
+    if bufnr ~= nil then
+        if bufnr == 0 then
+            bufnr = api.nvim_get_current_buf()
+        end
+        if M.enabled[bufnr] == nil then
+            M.enabled[bufnr] = false
+        end
+        return M.enabled[bufnr]
+    end
+end
+
+local function set_enabled(bufnr, enabled)
+    if bufnr == 0 then
+        bufnr = api.nvim_get_current_buf()
+    end
+    M.enabled[bufnr] = enabled
 end
 
-local function hide(bufnr)
-    api.nvim_buf_clear_namespace(bufnr, M.state.ns, 0, -1)
+-- end_line is inclusive
+local function del_hints(bufnr, start_line, end_line)
+    for _, tuple in ipairs(api.nvim_buf_get_extmarks(bufnr, ns, { start_line, 0 }, { end_line, -1 }, {})) do
+        api.nvim_buf_del_extmark(bufnr, ns, tuple[1])
+    end
+end
+
+local function del_all_hints()
+    for bufnr, _ in pairs(M.enabled) do
+        api.nvim_buf_clear_namespace(bufnr, ns, 0, -1)
+    end
+    M.enabled = {}
 end
 
 local function handler(err, result, ctx)
-    if not err and result and M.state.enabled then
-        local bufnr = ctx.bufnr
+    local bufnr = ctx.bufnr
+    if not err and result and buf_enabled(bufnr) then
         if not api.nvim_buf_is_loaded(bufnr) then
             return
         end
 
-        hide(bufnr)
-
         local hints = result.inlayHints or {}
         local parsed = {}
         for _, value in ipairs(hints) do
             local pos = value.position
-            local line_str = tostring(pos.line)
+            local line = pos.line
 
-            if parsed[line_str] then
-                table.insert(parsed[line_str], value)
-                table.sort(parsed[line_str], function(a, b)
+            if parsed[line] then
+                table.insert(parsed[line], value)
+                table.sort(parsed[line], function(a, b)
                     return a.position.character < b.position.character
                 end)
             else
-                parsed[line_str] = {
+                parsed[line] = {
                     value,
                 }
             end
         end
 
-        local lines = api.nvim_buf_get_lines(bufnr, 0, -1, false)
-        for key, value in pairs(parsed) do
-            local line = tonumber(key)
-            if lines[line + 1] then
+        local lines_cnt = vim.fn.line("$")
```

Comment:
```suggestion
        local lines_cnt = #api.nvim_buf_get_lines(bufnr, 0, -1, false)
```

Just a personal preference, but I prefer to avoid the Vim function, which is hard to understand if you're not familiar with Vimscript.

Reply:
Changed to `api.nvim_buf_line_count`. I didn't know this function exists while developing 😅